src/scripting_api/event.js

```diff
@@ -56,23 +71,115 @@ class EventDispatcher {
     const name = baseEvent.name.replace(" ", "");
     const source = this._objects[id];
     const event = (this._document.obj._event = new Event(baseEvent));
-    const oldValue = source.obj.value;
+    let savedChange;
+
+    if (source.obj._isButton()) {
+      source.obj._id = id;
+      event.value = source.obj._getExportValue(event.value);
+    }
+
+    if (name === "Keystroke") {
+      savedChange = {
+        value: event.value,
+        change: event.change,
+        selStart: event.selStart,
+        selEnd: event.selEnd,
+      };
+    } else if (name === "Blur" || name === "Focus") {
+      Object.defineProperty(event, "value", {
+        configurable: false,
+        writable: false,
+        enumerable: true,
+        value: event.value,
+      });
+    } else if (name === "Validate") {
+      this.runValidation(source, event);
+      return;
+    }
 
     this.runActions(source, source, event, name);
-    if (event.rc && oldValue !== event.value) {
-      source.wrapped.value = event.value;
+
+    if (name === "Keystroke") {
+      if (event.rc) {
+        if (event.willCommit) {
+          this.runValidation(source, event);
+        } else if (
+          event.change !== savedChange.change ||
+          event.selStart !== savedChange.selStart ||
+          event.selEnd !== savedChange.selEnd
+        ) {
+          source.wrapped.value = this.mergeChange(event);
+        }
+      } else if (!event.willCommit) {
+        source.obj._send({
+          id: source.obj._id,
+          value: savedChange.value,
+          selRange: [savedChange.selStart, savedChange.selEnd],
+        });
+      }
+    }
+  }
+
+  runValidation(source, event) {
+    const hasRan = this.runActions(source, source, event, "Validate");
+    if (event.rc) {
+      if (hasRan) {
+        source.wrapped.value = event.value;
+      } else {
+        source.obj.value = event.value;
+      }
+
+      if (this._document.obj.calculate) {
+        this.runCalculate(source, event);
+      }
+
+      event.value = source.obj.value;
+      this.runActions(source, source, event, "Format");
+      source.wrapped.valueAsString = event.value;
     }
   }
 
   runActions(source, target, event, eventName) {
     event.source = source.wrapped;
     event.target = target.wrapped;
     event.name = eventName;
+    event.targetName = target.obj.name;
     event.rc = true;
-    if (!target.obj._runActions(event)) {
-      return true;
+
+    return target.obj._runActions(event);
+  }
+
+  calculateNow() {
```

Comment:
Is this used?

Reply:
From here https://github.com/mozilla/pdf.js/blob/master/src/scripting_api/doc.js#L572.
and for the spec:
https://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/js_api_reference.pdf#page=277